users/models.py

```diff
@@ -1,3 +1,287 @@
 from django.db import models
+from django.contrib.auth.models import AbstractBaseUser, BaseUserManager, PermissionsMixin
+from django.core.validators import EmailValidator, RegexValidator
+from django.core.exceptions import ValidationError
+from django.utils import timezone
+from django.contrib.auth.password_validation import validate_password
+import logging
+import json
 
-# Create your models here.
+logger = logging.getLogger(__name__)
+
+
+class CustomUserManager(BaseUserManager):
+    """Custom user manager for the User model."""
+
+    def create_user(self, email, password=None, **extra_fields):
+        """
+        Create and save a new user with the given email and password.
+
+        Args:
+            email (str): User's email address.
+            password (str, optional): User's password.
+            **extra_fields: Additional fields for the user.
+
+        Returns:
+            User: Newly created user instance.
+
+        Raises:
+            ValueError: If email is not provided or password is invalid.
+        """
+        if not email:
+            raise ValueError('The Email field must be set')
+        email = self.normalize_email(email)
+        user = self.model(email=email, **extra_fields)
+        if password:
+            if not self.validate_password(password):
+                raise ValueError(
+                    'Password must be at least 8 characters long and contain at least one uppercase letter, one lowercase letter, one digit, and one special character.')
+            try:
+                validate_password(password)
+                user.set_password(password)
+            except ValidationError as e:
+                raise ValueError(f"Invalid password: {', '.join(e.messages)}")
+        else:
+            raise ValueError('Password must be provided')
+        user.save(using=self._db)
+        logger.info(f"New user created: {email}")
+        return user
+
+    def create_superuser(self, email, password=None, **extra_fields):
+        """
+        Create and save a new superuser with the given email and password.
+
+        Args:
+            email (str): Superuser's email address.
+            password (str, optional): Superuser's password.
+            **extra_fields: Additional fields for the superuser.
+
+        Returns:
+            User: Newly created superuser instance.
+        """
+        extra_fields.setdefault('is_staff', True)
+        extra_fields.setdefault('is_superuser', True)
+        extra_fields.setdefault('is_active', True)
+        user = self.create_user(email, password, **extra_fields)
+        if 'Admin' not in self.ALLOWED_ROLES:
+            raise ValidationError(f"Cannot add role: Admin")
+        user.add_role('Admin')
+        logger.info(f"New superuser created: {email}")
+        return user
```

Comment:
The method attempts to manually check if the `Admin` role is allowed. This logic should likely be handled more dynamically or through the role assignment functions rather than being hardcoded in the superuser creation logic.
Set default admin-related fields (`is_staff`, `is_superuser`) without manually checking if the `Admin` role exists. This would prevent unexpected failures if the role management logic changes later.

Reply:
done